core/src/main/scala/org/apache/spark/sql/execution/columnar/SampleColumnTableScan.scala

```diff
@@ -0,0 +1,669 @@
+      case GreaterThan(l: Literal, a: AttributeReference) => statsFor(a).lowerBound < l
+
+      case GreaterThanOrEqual(a: AttributeReference, l: Literal) => l <= statsFor(a).upperBound
+      case GreaterThanOrEqual(l: Literal, a: AttributeReference) => statsFor(a).lowerBound <= l
+
+      case IsNull(a: Attribute) => statsFor(a).nullCount > 0
+      case IsNotNull(a: Attribute) => statsFor(a).count - statsFor(a).nullCount > 0
+    }
+
+    // This code is picked up from InMemoryTableScanExec
+    val columnBatchStatFilters: Seq[Expression] = {
+      if (relation.isInstanceOf[BaseColumnFormatRelation]) {
+        allFilters.flatMap { p =>
+          val filter = buildFilter.lift(p)
+          val boundFilter = filter.map(BindReferences.bindReference(
+            _, columnBatchStatistics.schema, allowFailures = true))
+
+          boundFilter.foreach(_ =>
+            filter.foreach(f =>
+              logDebug(s"Predicate $p generates partition filter: $f")))
+
+          // If the filter can't be resolved then we are missing required statistics.
+          boundFilter.filter(_.resolved)
+        }
+      } else {
+        Seq.empty[Expression]
+      }
+    }
+
+    val columnBatchStatsSchema = getColumnBatchStatSchema
+    val numStatFields = columnBatchStatsSchema.length
+    val predicate = ExpressionCanonicalizer.execute(
+      BindReferences.bindReference(columnBatchStatFilters
+        .reduceOption(And).getOrElse(Literal(true)), columnBatchStatsSchema))
+    val statRow = ctx.freshName("statRow")
+    val statBytes = ctx.freshName("statBytes")
+    ctx.INPUT_ROW = statRow
+    ctx.currentVars = null
+    val predicateEval = predicate.genCode(ctx)
+
+    val platformClass = classOf[Platform].getName
+    val columnBatchesSeen = metricTerm(ctx, "columnBatchesSeen")
+    val columnBatchesSkipped = metricTerm(ctx, "columnBatchesSkipped")
+    val filterFunction = ctx.freshName("columnBatchFilter")
+    ctx.addNewFunction(filterFunction,
+      s"""
+         |private boolean $filterFunction(byte[] $statBytes) {
+         |  final UnsafeRow $statRow = new UnsafeRow($numStatFields);
+         |  $statRow.pointTo($statBytes, $platformClass.BYTE_ARRAY_OFFSET,
+         |    $statBytes.length);
+         |  $columnBatchesSeen.add(1);
+         |  // Skip the column batches based on the predicate
+         |  ${predicateEval.code}
+         |  if (!${predicateEval.isNull} && ${predicateEval.value}) {
+         |    return true;
+         |  } else {
+         |    $columnBatchesSkipped.add(1);
+         |    return false;
+         |  }
+         |}
+       """.stripMargin)
+    filterFunction
+  }
+
+  private val allRDDs = if (otherRDDs.isEmpty) rdd
+  else new UnionScanRDD(rdd.sparkContext, (Seq(rdd) ++ otherRDDs)
+    .asInstanceOf[Seq[RDD[Any]]])
+
+  private val isOffHeap = GemFireXDUtils.getGemFireContainer(
+    baseRelation.table, true).isOffHeap
+
+  private val otherRDDsPartitionIndex = rdd.getNumPartitions
```

Comment:
Since this file is a copy, lot of code is not making sense for reading. I would suggest a cleanup, and retaining only relevant code.